src/containerapp/azext_containerapp/_utils.py

```diff
@@ -12,145 +12,85 @@
-                         LOG_ANALYTICS_RP)
+                         LOG_ANALYTICS_RP, CONTAINER_APPS_RP, CHECK_CERTIFICATE_NAME_AVAILABILITY_TYPE)
+from ._models import (ContainerAppCustomDomainEnvelope as ContainerAppCustomDomainEnvelopeModel)
 
 logger = get_logger(__name__)
 
 
+def register_provider_if_needed(cmd, rp_name):
+    if not _is_resource_provider_registered(cmd, rp_name):
+        _register_resource_provider(cmd, rp_name)
+
+
 def validate_container_app_name(name):
     if name and len(name) > MAXIMUM_CONTAINER_APP_NAME_LENGTH:
         raise ValidationError(f"Container App names cannot be longer than {MAXIMUM_CONTAINER_APP_NAME_LENGTH}. "
                               f"Please shorten {name}")
 
 
-# original implementation at azure.cli.command_modules.role.custom.create_service_principal_for_rbac
-# reimplemented to remove incorrect warning statements
-def create_service_principal_for_rbac(  # pylint:disable=too-many-statements,too-many-locals, too-many-branches, unused-argument, inconsistent-return-statements
-        cmd, name=None, years=None, create_cert=False, cert=None, scopes=None, role=None,
-        show_auth_for_sdk=None, skip_assignment=False, keyvault=None):
-    from azure.cli.command_modules.role.custom import (_graph_client_factory, TZ_UTC, _process_service_principal_creds,
-                                                       _validate_app_dates, create_application,
-                                                       _create_service_principal, _create_role_assignment,
-                                                       _error_caused_by_role_assignment_exists)
-
-    if role and not scopes or not role and scopes:
-        raise ArgumentUsageError("Usage error: To create role assignments, specify both --role and --scopes.")
-
-    graph_client = _graph_client_factory(cmd.cli_ctx)
-
-    years = years or 1
-    _RETRY_TIMES = 36
-    existing_sps = None
+def retry_until_success(operation, err_txt, retry_limit, *args, **kwargs):
+    try:
+        return operation(*args, **kwargs)
+    except Exception as e:
+        retry_limit -= 1
+        if retry_limit <= 0:
+            raise CLIInternalError(err_txt) from e
+        time.sleep(5)
+        logger.info(f"Encountered error: {e}. Retrying...")
+
+
+def get_vnet_location(cmd, subnet_resource_id):
+    parsed_rid = parse_resource_id(subnet_resource_id)
+    vnet_client = network_client_factory(cmd.cli_ctx)
+    location = vnet_client.virtual_networks.get(resource_group_name=parsed_rid.get("resource_group"),
+                                                virtual_network_name=parsed_rid.get("name")).location
+    return _normalize_location(cmd, location)
+
+
+def create_service_principal_for_github_action(cmd, scopes=None, role="contributor"):
+    from azure.cli.command_modules.role.custom import (create_application, create_service_principal,
```

Comment:
Methods like `create_application` are for `role` modules internal private use and don't maintain its backward compatibility. Is it possible to directly use `GraphClient` retrieved from `azure.cli.command_modules.role.graph_client_factory`?

Reply:
My bad, I assumed that these functions were safe to use since they're tied to CLI commands (`az ad sp app create` etc.) and appeared to be public. 

I just removed the dependencies on all functions in `azure.cli.command_modules.role.custom`. Our extension is now using the SDK/client for all of these instead

Reply:
Just to clarify, this change is also in the latest commit I pushed